server/sonar-server/src/main/java/org/sonar/server/rule/index/RuleIndexer.java

```diff
@@ -19,78 +19,118 @@
  */
 package org.sonar.server.rule.index;
 
-import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import java.util.Arrays;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Set;
 import org.elasticsearch.action.index.IndexRequest;
 import org.sonar.api.rule.RuleKey;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
 import org.sonar.db.organization.OrganizationDto;
 import org.sonar.server.es.BulkIndexer;
 import org.sonar.server.es.BulkIndexer.Size;
 import org.sonar.server.es.EsClient;
+import org.sonar.server.es.IndexType;
+import org.sonar.server.es.StartupIndexer;
 
+import static java.util.Collections.singletonList;
 import static org.sonar.server.rule.index.RuleIndexDefinition.INDEX_TYPE_RULE;
+import static org.sonar.server.rule.index.RuleIndexDefinition.INDEX_TYPE_RULE_EXTENSION;
 
-public class RuleIndexer {
+public class RuleIndexer implements StartupIndexer {
 
   private final EsClient esClient;
-  private final RuleIteratorFactory ruleIteratorFactory;
+  private final DbClient dbClient;
 
-  public RuleIndexer(EsClient esClient, RuleIteratorFactory ruleIteratorFactory) {
+  public RuleIndexer(EsClient esClient, DbClient dbClient) {
     this.esClient = esClient;
-    this.ruleIteratorFactory = ruleIteratorFactory;
+    this.dbClient = dbClient;
   }
 
-  public void index(OrganizationDto organization, RuleKey ruleKey) {
-    BulkIndexer bulk = createBulkIndexer(Size.REGULAR);
-    try (RuleIterator rules = ruleIteratorFactory.createForKey(organization, ruleKey)) {
-      doIndex(bulk, rules);
-    }
+  @Override
+  public Set<IndexType> getIndexTypes() {
+    return ImmutableSet.of(INDEX_TYPE_RULE, INDEX_TYPE_RULE_EXTENSION);
   }
 
-  public void index(OrganizationDto organization, List<RuleKey> ruleKeys) {
-    BulkIndexer bulk = createBulkIndexer(Size.REGULAR);
-    try (RuleIterator rules = ruleIteratorFactory.createForKeys(organization, ruleKeys)) {
-      doIndex(bulk, rules);
+  @Override
+  public void indexOnStartup(Set<IndexType> emptyIndexTypes) {
```

Comment:
In fact the term is not correct. It should be "uninitializedIndexTypes". The fact that it's empty is a consequence.